File: DiskAccess.fs
module DiskAccess
open FSharpTools
open Configuration
open Option

let private switchDiskOff port = 
    async {
        printfn "switching disk off..."
        let! result = Process.asyncRunCmd "/usr/sbin/uhubctl" <| (sprintf "-l 1-1 -a 0 -p %d -r 500" port)
        printfn "disk switched off\n%s" result
    } |> Async.Start

let private createTimer () =
    let createTimer port = 
        let timer = new System.Timers.Timer 300_000
        timer.AutoReset <- false
        timer.Elapsed.Add (fun _ -> switchDiskOff port) 
        timer, port

    getUsbPort ()
    |> map createTimer

let private shutdownTimer = createTimer ()

let access () =
    let access (timer: System.Timers.Timer, port) =
        async {
            printfn "accessing disk..."
            let! result = Process.asyncRunCmd "/usr/sbin/uhubctl" <| (sprintf "-l 1-1 -a 1 -p %d" port)
            do! Async.Sleep(6000)
            let! mountResult =  Process.asyncRunCmd "/usr/bin/mount" "-a"

            if timer.Enabled then
                timer.Stop ()
            timer.Start ()
            printfn "%s\n%s" result mountResult
        }

    async {
        do! 
            shutdownTimer 
            |> iterAsync access 
    }

let needed () =
    let needed (timer: System.Timers.Timer, _) = 
        if timer.Enabled then
            timer.Stop ()
            timer.Start ()

    shutdownTimer
    |> iter needed

File: DiskAccess.cs
using static CsTools.ProcessCmd;
using static System.Console;
using static Configuration;
using static CsTools.Core;
using CsTools.Extensions;

static class DiskAccess
{
    public static async Task AccessDisk(HttpContext context)
    {
        await AccessDisk();
        await context.Response.WriteAsync("Disk accessed");
    }

    public static Task DiskNeeded(HttpContext context)
        => context
            .SideEffect(_ => 
            {
                WriteLine("Disk needed");
                if (diskShutdownTimer.Enabled) {
                    diskShutdownTimer.Stop();
                    diskShutdownTimer.Start();
                } 
                else
                    StartAccessDisk();
            })
            .Response
            .WriteAsync("Disk shutdown delayed");

    static async Task AccessDisk()
    {
        WriteLine("Accessing disk");
        var result = await Try(() =>  RunAsync("/usr/sbin/uhubctl", $"-l 1-1 -a 1 -p {GetEnvironmentVariable(UsbMediaPort)}"), e => $"{e}");
        await Task.Delay(6000);
        var mountResult = await Try(() =>  RunAsync("/usr/bin/mount", "-a"), e => $"{e}");
        if (diskShutdownTimer.Enabled)
            diskShutdownTimer.Stop();
        diskShutdownTimer.Start();
        WriteLine(result);
        WriteLine(mountResult);
    }

    static async void StartAccessDisk()
    {
        try 
        {
            await AccessDisk();
        }
        catch {}
    }
    static DiskAccess()
        => diskShutdownTimer = new System.Timers.Timer(300_000)
        {
            AutoReset = false,
        }.SideEffect(t => t.Elapsed += async (s, e) =>
        {
            WriteLine("Switching disk off...");
            var result = await Try(() =>  RunAsync("/usr/sbin/uhubctl", $"-l 1-1 -a 0 -p {GetEnvironmentVariable(UsbMediaPort)} -r 500"), e => $"{e}");
            WriteLine("disk switched off");
            WriteLine(result);
            diskShutdownTimer?.Stop();
        });

    static System.Timers.Timer diskShutdownTimer;
}